### PR TITLE
Add a/b test for implicit follow points

### PR DESCRIPTION
**Pull request by benhalpern**
## What type of PR is this? (check all applicable)

- [ ] Refactor
- [ ] Feature
- [ ] Bug Fix
- [x] Optimization
- [ ] Documentation Update

## Description

This adds an a/b test to gather information on the relative utility gained from different implementations of calculating "implicit points" for tag following.

I also added general-purpose docs about a/b testing within the codebase.

## Related Tickets & Documents

https://github.com/forem/InternalProjectPlanning/issues/373

## Added tests?

- [x] Yes
- [ ] No, and this is why: _please replace this line with details on why tests
      have not been included_
- [ ] I need help with writing tests

## Added to documentation?

- [x] [Developer Docs](https://docs.forem.com) and/or [Admin Guide](https://forem.gitbook.io/forem-admin-guide/)
- [ ] README
- [ ] No documentation needed

## [optional] What gif best describes this PR or how it makes you feel?

doc brown

### CODE REVIEW

File: docs/technical-overview/ab_testing.md [hunk 1]
```diff
@@ -0,0 +1,7 @@
+# A/B testing
+
+We use the [Field Test](https://github.com/ankane/field_test) gem for conducting simple a/b tests.
+
+If you want to propose an a/b test of a feature, you may make a pull request which defines the hypotheses and what admins should look for to declare a winner. As a/b tests are going to have results that may differ from Forem to Forem, the process is relatively immature. In the future we may have more studies that can return anonymous ecosystem-wide results.
+
+A/B test are inherently most useful in _large_ Forems, where qualitative feedback may be more useful on small Forems. As such, [DEV](https://dev.to) is the most important Forem for some A/B tests, with the caviat that DEV results may not apply well to future large Forems. We should seek to re-run useful experiments within the ecosystem after time has passed.
```
Comment (mstruve) [inline docs/technical-overview/ab_testing.md:7]:
```suggestion
A/B tests are inherently the most useful in _large_ Forems, where qualitative feedback may be more useful on small Forems. As such, [DEV](https://dev.to) is our largest Forem and therefore can provide us with most feedback for our existing A/B tests. However, we must keep in mind that DEV results may not apply well to future large Forems. We should seek to re-run useful experiments within the ecosystem after time has passed.
```

File: config/field_test.yml [hunk 2]
```diff
@@ -9,6 +9,30 @@ experiments:
       - user_creates_reaction
       - user_views_article_four_days_in_week
       - user_views_article_four_hours_in_day
+  follow_implicit_points: # Points given for implicit tag weights
+    # Implemented here: app/workers/follows/update_points_worker.rb
+    # Hypotheses: Weighting tags based on implied preference will lead
+    # to more repeat engagement.
+    # Recommended duration: 4 weeks
```
Comment (mstruve) [inline config/field_test.yml:16]: ❤️  Maybe put the date we are starting this so we don't have to go to the git blame to check when this comment was added.

File: app/workers/follows/update_points_worker.rb [hunk 3]
```diff
@@ -34,7 +35,8 @@ def calculate_implicit_points(tag, user)
       tags = articles.pluck(:cached_tag_list).flat_map { |list| list.split(", ") }
       occurrences = tags.count(tag.name)
       bonus = inverse_popularity_bonus(tag)
-      Math.log(occurrences + bonus + 1) # +1 is purelt to avoid log(0) => -infinity
+      finalized_points(occurrences, bonus, user)
+      # Math.log(occurrences + bonus + 1) # +1 is purelt to avoid log(0) => -infinity
```
Comment (mstruve) [inline app/workers/follows/update_points_worker.rb:39]: I think we can remove this line and if you want add a comment below to the "base" explaining that's our existing implementation.

**Review by mstruve — COMMENTED** [review]
Couple cleanup suggestions otherwise this looks great!!! Thanks for adding it. 

Also, should we remove the `user_home_feed` test from the yml file? I probably should have done that.

**Review by mstruve — APPROVED** [review]
🚀

File: docs/technical-overview/ab_testing.md [hunk 4]
```diff
@@ -0,0 +1,7 @@
+# A/B testing
+
+We use the [Field Test](https://github.com/ankane/field_test) gem for conducting simple a/b tests.
+
+If you want to propose an a/b test of a feature, you may make a pull request which defines the hypotheses and what admins should look for to declare a winner. As a/b tests are going to have results that may differ from Forem to Forem, the process is relatively immature. In the future we may have more studies that can return anonymous ecosystem-wide results.
+
+A/B tests are inherently the most useful in _large_ Forems, where qualitative feedback may be more useful on small Forems. As such, [DEV](https://dev.to) is our largest Forem and therefore can provide us with most feedback for our existing A/B tests. However, we must keep in mind that DEV results may not apply well to future large Forems. We should seek to re-run useful experiments within the ecosystem after time has passed.
```
Comment (jacobherrington) [inline docs/technical-overview/ab_testing.md:7]:
```suggestion
A/B tests are inherently the most useful in _large_ Forems, where qualitative feedback may be more useful on small Forems. As such, [DEV](https://dev.to) is our largest Forem and therefore can provide us with the most feedback for our existing A/B tests. However, we must keep in mind that DEV results may not apply well to future large Forems. We should seek to re-run useful experiments within the ecosystem after time has passed.
```

**Review by jacobherrington — COMMENTED** [review]
Looks good, just one typo that I found. :+1:

File: config/field_test.yml (hunk 2 above)
Comment (benhalpern) [inline config/field_test.yml:16]:
```suggestion
    # Recommended duration: 4 weeks (ending at start of January)
```

**Review by joshpuetz — APPROVED** [review]
These are great additions, thanks!

File: docs/technical-overview/ab_testing.md [hunk 5]
```diff
@@ -0,0 +1,7 @@
+# A/B testing
+
+We use the [Field Test](https://github.com/ankane/field_test) gem for conducting simple a/b tests.
```
Comment (juliannatetreault) [inline docs/technical-overview/ab_testing.md:3]: _Quick, non-blocking nitpick:_
```suggestion
We use the [Field Test](https://github.com/ankane/field_test) gem for conducting simple A/B tests.
```

File: docs/technical-overview/ab_testing.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
...
+
+If you want to propose an a/b test of a feature, you may make a pull request which defines the hypotheses and what admins should look for to declare a winner. As a/b tests are going to have results that may differ from Forem to Forem, the process is relatively immature. In the future we may have more studies that can return anonymous ecosystem-wide results.
```
Comment (juliannatetreault) [inline docs/technical-overview/ab_testing.md:5]: _Another quick, non-blocking nitpick:_
```suggestion
If you want to propose an A/B test of a feature, you may make a pull request which defines the hypotheses and what admins should look for to declare a winner. As A/B tests are going to have results that may differ from Forem to Forem, the process is relatively immature. In the future we may have more studies that can return anonymous ecosystem-wide results.
```

**Review by juliannatetreault — APPROVED** [review]
Nothing seems to be broken and everything seems to look good. 👍 Thank you so much for all of the comments and the link to the `Field Test` gem - super helpful during review!

**Pull request merged**